【Hackathon 7th No.21】为 Paddle 新增 reset_max_memory_reserved/reset_max_memory_allocated API -part

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Deprecations | Others ] -->

New features

### Description
<!-- Describe what you’ve done -->

https://github.com/PaddlePaddle/community/blob/master/hackathon/hackathon_7th/%E3%80%90Hackathon%207th%E3%80%91%E4%B8%AA%E4%BA%BA%E6%8C%91%E6%88%98%E8%B5%9B%E2%80%94%E6%A1%86%E6%9E%B6%E5%BC%80%E5%8F%91%E4%BB%BB%E5%8A%A1%E5%90%88%E9%9B%86.md#no21-%E4%B8%BA-paddle-%E6%96%B0%E5%A2%9E-reset_peak_memory_statsreset_max_memory_allocatedmemory_stats-api

- RFC文档：https://github.com/PaddlePaddle/community/pull/1022
- API文档：https://github.com/PaddlePaddle/docs/pull/7013

Qin-sx committed:
added reset peak value initialization

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   python/paddle/device/cuda/__init__.py

Qin-sx committed:
added comments

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   python/paddle/device/cuda/__init__.py

Qin-sx committed:
added cpp tests

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   test/cpp/fluid/memory/stats_test.cc

Qin-sx committed:
added python tests

	new file:   test/legacy_test/test_cuda_memory_stats.py
	new file:   test/legacy_test/test_cuda_reset_peak_memory_stats.py

Qin-sx committed:
added a python test for reset_max_memory_allocated

	new file:   test/legacy_test/test_cuda_reset_max_memory_allocated.py

Qin-sx committed:
formatted by pre-commit

	modified:   python/paddle/device/cuda/__init__.py
	modified:   test/legacy_test/test_cuda_memory_stats.py
	modified:   test/legacy_test/test_cuda_reset_max_memory_allocated.py
	modified:   test/legacy_test/test_cuda_reset_peak_memory_stats.py

Qin-sx committed:
formatted by pre-commit (clang-format)
	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   test/cpp/fluid/memory/stats_test.cc

Qin-sx committed:
added reset max memory reserved function

	modified:   python/paddle/device/cuda/__init__.py
	modified:   test/legacy_test/test_cuda_reset_max_memory_allocated.py
	new file:   test/legacy_test/test_cuda_reset_max_memory_reserved.py

From00 commented on python/paddle/device/cuda/__init__.py:
```diff
@@ -298,6 +302,142 @@ def max_memory_reserved(device: _CudaPlaceLike | None = None) -> int:
+
+def reset_max_memory_allocated(device: _CudaPlaceLike | None = None) -> None:
+    '''
+    Reset the peak values of GPU memory allocated to the current values.
```
接口说明最好和max_memory_allocated保持一致。
`Reset the peak size of GPU memory that is held by the allocator of the given device.`

Qin-sx replied:
嗯，收到

From00 commented on test/cpp/fluid/memory/stats_test.cc:
```diff
@@ -94,6 +97,18 @@ class StatsTest : public ::testing::Test {
+
+      EXPECT_GE(peak_value_func_(stat_type_, 0),
+                current_value_func_(stat_type_, 0));
+      reset_peak_value_func_(stat_type_, 0);
```
在这个测试中，这一步如果reset不成功，是不是后面的检查也是会通过的？

Qin-sx replied:
如果reset不成功，下一行(107行)的检查应该无法通过。

```c++
  void ResetPeakValueTest() {
    for (int64_t data : datas_) {
      update_func_(stat_type_, 0, data);

      EXPECT_GE(peak_value_func_(stat_type_, 0),
                current_value_func_(stat_type_, 0));
      // reset_peak_value_func_(stat_type_, 0);
      printf("data: %ld, Peak Value: %ld, Current Value: %ld\n",data, peak_value_func_(stat_type_, 0), current_value_func_(stat_type_, 0));
      EXPECT_EQ(peak_value_func_(stat_type_, 0),
                current_value_func_(stat_type_, 0));
    }
  }
```

如果将`reset_peak_value_func_`函数注释，测试将无法通过。下面为部分运行结果。

```
115: Test timeout computed to be: 10000000
115: [==========] Running 4 tests from 1 test case.
115: [----------] Global test environment set-up.
115: [----------] 4 tests from StatsTest
115: [ RUN      ] StatsTest.DeviceAllocatedTest
115: data: 543149808935355, Peak Value: 45703145873829393, Current Value: 45703145873829393
115: data: 634698327471328, Peak Value: 46337844201300721, Current Value: 46337844201300721
115: data: 706215795436611, Peak Value: 47044059996737332, Current Value: 47044059996737332
115: data: 577939367795333, Peak Value: 47621999364532665, Current Value: 47621999364532665
115: data: 419479490054362, Peak Value: 48041478854587027, Current Value: 48041478854587027
115: data: 21975227714595, Peak Value: 48063454082301622, Current Value: 48063454082301622
115: data: 812939817942250, Peak Value: 48876393900243872, Current Value: 48876393900243872
115: data: 984428837942082, Peak Value: 49860822738185954, Current Value: 49860822738185954
115: data: 537304104446806, Peak Value: 50398126842632760, Current Value: 50398126842632760
115: data: 685008544452453, Peak Value: 51083135387085213, Current Value: 51083135387085213
115: data: 563352858161268, Peak Value: 51646488245246481, Current Value: 51646488245246481
115: data: 690143831596330, Peak Value: 52336632076842811, Current Value: 52336632076842811
115: data: 964829938186077, Peak Value: 53301462015028888, Current Value: 53301462015028888
115: data: 476984078018245, Peak Value: 53778446093047133, Current Value: 53778446093047133
115: data: 804403365180177, Peak Value: 54582849458227310, Current Value: 54582849458227310
115: data: -57918691189304, Peak Value: 54582849458227310, Current Value: 54524930767038006
115: /home/aistudio/test/Paddle/test/cpp/fluid/memory/stats_test.cc:109: Failure
115: Expected equality of these values:
115:   peak_value_func_(stat_type_, 0)
115:     Which is: 54582849458227310
115:   current_value_func_(stat_type_, 0)
115:     Which is: 54524930767038006
115: data: 947611269236893, Peak Value: 55472542036274899, Current Value: 55472542036274899
115: data: 752188963801927, Peak Value: 56224731000076826, Current Value: 56224731000076826
115: data: 710946451346683, Peak Value: 56935677451423509, Current Value: 56935677451423509
115: data: -49226452527666, Peak Value: 56935677451423509, Current Value: 56886450998895843
115: /home/aistudio/test/Paddle/test/cpp/fluid/memory/stats_test.cc:109: Failure
115: Expected equality of these values:
115:   peak_value_func_(stat_type_, 0)
115:     Which is: 56935677451423509
115:   current_value_func_(stat_type_, 0)
115:     Which is: 56886450998895843
115: data: -59049377393968, Peak Value: 56935677451423509, Current Value: 56827401621501875
115: /home/aistudio/test/Paddle/test/cpp/fluid/memory/stats_test.cc:109: Failure
115: Expected equality of these values:
115:   peak_value_func_(stat_type_, 0)
115:     Which is: 56935677451423509
115:   current_value_func_(stat_type_, 0)
115:     Which is: 56827401621501875
115: data: 14128239868858, Peak Value: 56935677451423509, Current Value: 56841529861370733
115: /home/aistudio/test/Paddle/test/cpp/fluid/memory/stats_test.cc:109: Failure
115: Expected equality of these values:
115:   peak_value_func_(stat_type_, 0)
115:     Which is: 56935677451423509
115:   current_value_func_(stat_type_, 0)
115:     Which is: 56841529861370733
115: data: 463298869064035, Peak Value: 57304828730434768, Current Value: 57304828730434768
```

Qin-sx committed:
deleted memory stats and reset peak memory stats

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   python/paddle/device/cuda/__init__.py
	deleted:    test/legacy_test/test_cuda_memory_stats.py
	deleted:    test/legacy_test/test_cuda_reset_peak_memory_stats.py

From00 reviewed (approved):
LGTM

luotao1 commented:
<img width="725" alt="image" src="https://github.com/user-attachments/assets/865bc95d-a8d6-46e2-adfb-707ac55104df" />
覆盖率不够，需要加单测

SigureMo reviewed (approved):
LGTMeow <img src="https://www.gstatic.com/android/keyboard/emojikitchen/20231113/u1f4f0/u1f4f0_u1f43e.png" width="14" alt="🐾"/> for new pybind API without type annotations

Qin-sx commented:
> <img alt="image" width="725" src="https://private-user-images.githubusercontent.com/6836917/395985360-865bc95d-a8d6-46e2-adfb-707ac55104df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMzk4MDIsIm5iZiI6MTczNDMzOTUwMiwicGF0aCI6Ii82ODM2OTE3LzM5NTk4NTM2MC04NjViYzk1ZC1hOGQ2LTQ2ZTItYWRmYi03MDdhYzU1MTA0ZGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMDg1ODIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDc3MDQ4ZDc1MzEzZmEwMzAxYWUyMDk5NzFkOGE0OWQ0NjcwMGJmNTAyMDZmYWExYmVjZGI2MjU5YzlkZGFhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gss2FqI0d1ZDbKGVlu84UOD6vvIuswmrW6I4ql80t3c"> 覆盖率不够，需要加单测

未覆盖的代码为`Paddle/python/paddle/device/cuda/__init__.py`文件中的326和356行。当Paddle由CPU编译时，这两行会运行。  
![Lcov](https://github.com/user-attachments/assets/c9fd68d1-1080-4d07-87a5-6d58db237e65)

在测试文件`Paddle/test/legacy_test/test_cuda_reset_max_memory_allocated.py`和`backup/Paddle/test/legacy_test/test_cuda_reset_max_memory_reserved.py`中的85行有对此情况的测试。
```python
    def test_reset_max_memory_allocated_exception(self):
        if core.is_compiled_with_cuda():
            wrong_device = [
                core.CPUPlace(),
                device_count() + 1,
                -2,
                0.5,
                "gpu1",
            ]
            for device in wrong_device:
                with self.assertRaises(BaseException):  # noqa: B017
                    reset_max_memory_allocated(device)
        else:
            with self.assertRaises(ValueError):
                reset_max_memory_allocated()

    def test_reset_max_memory_reserved_exception(self):
        if core.is_compiled_with_cuda():
            wrong_device = [
                core.CPUPlace(),
                device_count() + 1,
                -2,
                0.5,
                "gpu1",
            ]
            for device in wrong_device:
                with self.assertRaises(BaseException):  # noqa: B017
                    reset_max_memory_reserved(device)
        else:
            with self.assertRaises(ValueError):
                reset_max_memory_reserved()
```

当由CPU编译的Paddle执行测试时，上述两行代码会被覆盖到。

jeff41404 commented:
please add link of rfc and PR of chinese document in description above

luotao1 commented:
please add link of rfc and PR of chinese document in description above

@jeff41404 Done

jeff41404 reviewed (approved):
LGTM

sunzhongkai588 reviewed (approved):
LGTM for docs

luotao1 commented:
hi, @Qin-sx 
- 非常感谢你对飞桨的贡献，我们正在运营一个PFCC组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨做贡献，详情可见 https://github.com/luotao1 主页说明。
- 如果你对PFCC有兴趣，请发送邮件至 [ext_paddle_oss@baidu.com](mailto:ext_paddle_oss@baidu.com)，我们会邀请你加入~
